### ISSUE TITLE
Exceptions are expensive, use Operation Result pattern instead

### ISSUE THREAD

**Issue by devbrsa** [GFlisch/Arc4u.Guidance.Doc#28]
https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/exceptions-and-performance?redirectedfrom=MSDN

Why not?
LanguageExt.Core - [LanguageExt.Core](https://github.com/louthy/language-ext)

**Comment by vvdb-architecture** [comment]
I agree we should apply Tester-Doer and TryParse patterns as much as possible, and I have seen code (even in Arc4u) where their application is not as efficient as can be. For example (<https://github.com/GFlisch/Arc4u/blob/master/src/Arc4u.Standard.Caching/CacheContext.cs>):

~~~csharp
// look inside the collection if the cache is initialized.
if (_caches.ContainsKey(cacheName))
    return _caches[cacheName];:
~~~

This is very inefficient since it results in two lookups: one for `ContainsKey` (whose result is thrown away because just a boolean is returned) and another one for `_caches[cacheName]`.

It is much better to write:

~~~csharp
// look inside the collection if the cache is initialized.
if (_caches.TryGetValue(cacheName, out var cache))
    return cache;
~~~

This is exactly the same and twice as efficient. In hot paths, it makes a difference.

This being said, exceptions, as their name indicates, are supposed _exceptional_. They are never supposed to be on a hot path, so their overhead (because they indeed have non-negligible overhead) should never be the dominant factor in code execution.

Functional programming is interesting and LanguageExt.Core is very cool, but we should be very careful when we consider introducing yet another dependency for marginal gains. It also would require familiarity with the dependency for every developer maintaining the code.

Also, I am a big coward and reading things like this (https://github.com/louthy/language-ext):

> This library uses and abuses the features of C# to provide a functional-programming 'base class library' that, if you squint, can look like extensions to the language itself. 


... makes me very, very afraid. Use and abuse: yikes!

**Comment by devbrsa** [comment]
Enter topics not explanation.
Was targeting the Exception thrown instead of returning them instead of throw.

https://medium.com/@cummingsi1993/the-operation-result-pattern-a-simple-guide-fe10ff959080

**Comment by rdarko** [comment]
Related to the #103

**Comment by devbrsa** [comment]
Exceptions are considered expensive in .NET because they involve creating an exception object, which takes up memory and processing power. Additionally, when an exception is thrown, the program's execution flow jumps to the nearest catch block, which can cause additional performance overhead.

The operation result pattern is a way to handle errors and return values in a way that avoids throwing exceptions. Instead of throwing exceptions, methods that use the operation result pattern return an object that contains both the result of the operation (if successful) and any error information (if the operation failed). This pattern can be more efficient than throwing exceptions because it avoids the overhead of creating and handling exception objects, and it allows the calling code to handle errors in a more controlled and predictable way. Additionally, it allows to handle errors in a more elegant way and can help to avoid using try-catch blocks everywhere in the code.

Basically exceptions should be thrown by internal/infrastructure components, if the issue is known is not an exception.
